### PR TITLE
Add DNS Load Balance for Server Domain

### PR DESCRIPTION
**Pull request by yyk123**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

- [x] [Searched](https://github.com/shadowsocks/shadowsocks-windows/search?q=is%3Apr&type=Issues) for similar pull requests
- [x] Compiled the code with Visual Studio

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New feature

---

### Description of your *pull request* and other information
If a domain name has more than 1 IP in DNS, rotate the returned IP list to pick up different IP.
Add DNS Load Balance and test available ip address for Server Domain if Server domain name have multiple IP resolved.

### CODE REVIEW

**Comment by celeron533** [comment]
`.github/PULL_REQUEST_TEMPLATE.md` is the template file. Please do not modify it.

**Comment by yyk123** [comment]
> `.github/PULL_REQUEST_TEMPLATE.md` is the template file. Please do not modify it.

Yeah, I revised it.

File: shadowsocks-csharp/Util/Sockets/SocketUtil.cs [hunk 1]
```diff
@@ -14,14 +13,13 @@ public class HostInfo
...
-        private Timer timer;
+        private System.Timers.Timer timer;
         private DateTime dateTime;
-
         public IPAddress IP
         {
             get
             {
-                return GetNextIP();
+                return GetNextIP();
```
Comment (pexcn) [inline shadowsocks-csharp/Util/Sockets/SocketUtil.cs]: 这里多出几个空格了

File: shadowsocks-csharp/Controller/Service/TCPRelay.cs [hunk 2]
```diff
@@ -566,12 +566,28 @@ private void StartConnect()
             try
             {
                 CreateRemote();
-
+                Server server=new Server(_server);
```
Comment (celeron533) [inline shadowsocks-csharp/Controller/Service/TCPRelay.cs]: MemberwiseClone could be another option. (No worries, no need to change current code. It's fine.)

**Comment by celeron533** [comment]
My concerns:
1. Switching between different server IP may cause some application malfunction, especially online gaming. Unless your server works like a CDN (multiple shadowsocks listeners on different IPs but only one IP is used to connect to the "real Internet")
2. Interval ping the server addresses could be a kind of fingerprint.

**Comment by yyk123** [comment]
> My concerns:
> 
> 1. Switching between different server IP may cause some application malfunction, especially online gaming. Unless your server works like a CDN (multiple shadowsocks listeners on different IPs but only one IP is used to connect to the "real Internet")
> 2. Interval ping the server addresses could be a kind of fingerprint.

And if someone wants to use it for some applications, I think it is better to use real ip rather than domains.  It also works fine if they bind just one ip on their domain.

I wrote this code to make it easier for certain people.
**Once they configured, there is no need to change anymore. **

 **The current version also cannot return a specific IP address, if someone binds multiple IP addresses on his domain.**
So, there is no change if some one use real IP address or bind just one IP address on the domain.

**As you mentioned, it indeed can be a fingerprint if we ping our vps by fixed interval. Maybe it is better to use random value between minimum  and maximum (e.g. (10min, 120min)). Or just delete it, so the software must be restarted to find the available IP (ping one time).** 
By the way, I think this feature won't be used by  too many people, because they must purchase one domain and bind ip on it.

As for me, it is very convenient. I do not need change my configuration once my vps had been banned. What I need to do , is do nothing (i.e. I have muliple IP) or change another ip or just delete the useless ip on my domain.

**Comment by yyk123** [comment]
checks failed caused by RNG.cs Close() multiple thread access _rng,  another thread set  _rng =null

**Comment by celeron533** [comment]
Hi @yyk123 , I found something interesting.
https://stackoverflow.com/questions/10257969/is-it-possible-that-one-domain-name-has-multiple-corresponding-ip-addresses
https://en.wikipedia.org/wiki/Round-robin_DNS

Tested on my local environment. You can see that each time the 1st returned IP is different.
```
C:\Users\********>nslookup amazon.com
Server:  *****************
Address:  ***.***.***.***

Non-authoritative answer:
Name:    amazon.com
Addresses:  176.32.103.205
          176.32.98.166
          205.251.242.103


C:\Users\********>nslookup amazon.com
Server:  *****************
Address:  ***.***.***.***

Non-authoritative answer:
Name:    amazon.com
Addresses:  176.32.98.166
          205.251.242.103
          176.32.103.205


C:\Users\********>nslookup amazon.com
Server:  *****************
Address:  ***.***.***.***

Non-authoritative answer:
Name:    amazon.com
Addresses:  205.251.242.103
          176.32.103.205
          176.32.98.166


C:\Users\********>nslookup amazon.com
Server:  *****************
Address:  ***.***.***.***

Non-authoritative answer:
Name:    amazon.com
Addresses:  176.32.103.205
          176.32.98.166
          205.251.242.103
```

```
C:\Users\********>nslookup bing.com
Server:  *****************
Address:  ***.***.***.***

Non-authoritative answer:
Name:    bing.com
Addresses:  2620:1ec:c11::200
          13.107.21.200
          204.79.197.200


C:\Users\********>nslookup bing.com
Server:  *****************
Address:  ***.***.***.***

Non-authoritative answer:
Name:    bing.com
Addresses:  2620:1ec:c11::200
          204.79.197.200
          13.107.21.200


C:\Users\********>nslookup bing.com
Server:  *****************
Address:  ***.***.***.***

Non-authoritative answer:
Name:    bing.com
Addresses:  2620:1ec:c11::200
          13.107.21.200
          204.79.197.200
```

**Comment by yyk123** [comment]
> Hi @yyk123 , I found something interesting.
> https://stackoverflow.com/questions/10257969/is-it-possible-that-one-domain-name-has-multiple-corresponding-ip-addresses
> https://en.wikipedia.org/wiki/Round-robin_DNS
> 
> Tested on my local environment. You can see that each time the 1st returned IP is different.
> 
> ```
> C:\Users\********>nslookup amazon.com
> Server:  *****************
> Address:  ***.***.***.***
> 
> Non-authoritative answer:
> Name:    amazon.com
> Addresses:  176.32.103.205
>           176.32.98.166
>           205.251.242.103
> 
> 
> C:\Users\********>nslookup amazon.com
> Server:  *****************
> Address:  ***.***.***.***
> 
> Non-authoritative answer:
> Name:    amazon.com
> Addresses:  176.32.98.166
>           205.251.242.103
>           176.32.103.205
> 
> 
> C:\Users\********>nslookup amazon.com
> Server:  *****************
> Address:  ***.***.***.***
> 
> Non-authoritative answer:
> Name:    amazon.com
> Addresses:  205.251.242.103
>           176.32.103.205
>           176.32.98.166
> 
> 
> C:\Users\********>nslookup amazon.com
> Server:  *****************
> Address:  ***.***.***.***
> 
> Non-authoritative answer:
> Name:    amazon.com
> Addresses:  176.32.103.205
>           176.32.98.166
>           205.251.242.103
> ```
> 
> ```
> C:\Users\********>nslookup bing.com
> Server:  *****************
> Address:  ***.***.***.***
> 
> Non-authoritative answer:
> Name:    bing.com
> Addresses:  2620:1ec:c11::200
>           13.107.21.200
>           204.79.197.200
> 
> 
> C:\Users\********>nslookup bing.com
> Server:  *****************
> Address:  ***.***.***.***
> 
> Non-authoritative answer:
> Name:    bing.com
> Addresses:  2620:1ec:c11::200
>           204.79.197.200
>           13.107.21.200
> 
> 
> C:\Users\********>nslookup bing.com
> Server:  *****************
> Address:  ***.***.***.***
> 
> Non-authoritative answer:
> Name:    bing.com
> Addresses:  2620:1ec:c11::200
>           13.107.21.200
>           204.79.197.200
> ```

Yeah, the sequence of the return ips is not sure. But it doesn't matter, I send DNS request every 15 minutes and switch ip every 10 seconds.

**Pull request closed without merging**